src/coprocessor/dag/batch_executor/limit.rs

```diff
@@ -0,0 +1,216 @@
+// limitations under the License.
+
+use super::interface::*;
+use crate::coprocessor::Result;
+use tipb::expression::FieldType;
+
+pub struct BatchLimitExecutor<Src: BatchExecutor> {
+    src: Src,
+    expect_rows: usize,
+}
+
+impl<Src: BatchExecutor> BatchLimitExecutor<Src> {
+    pub fn new(src: Src, limit: usize) -> Result<Self> {
+        if limit == 0 {
+            return Err(box_err!("limit should not be zero"));
+        }
+        Ok(Self {
+            src,
+            expect_rows: limit,
+        })
+    }
+}
+
+impl<Src: BatchExecutor> BatchExecutor for BatchLimitExecutor<Src> {
+    #[inline]
+    fn next_batch(&mut self, expect_rows: usize) -> BatchExecuteResult {
+        let mut result = self.src.next_batch(expect_rows);
+        if result.data.rows_len() <= self.expect_rows {
+            self.expect_rows -= result.data.rows_len();
+            return result;
+        }
+
+        result.data.truncate(self.expect_rows);
+        self.expect_rows = 0;
+        result.is_drained = Ok(true);
+        result
+    }
+
+    #[inline]
+    fn collect_statistics(&mut self, destination: &mut BatchExecuteStatistics) {
+        self.src.collect_statistics(destination);
+    }
+
+    #[inline]
+    fn schema(&self) -> &[FieldType] {
+        self.src.schema()
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::coprocessor::codec::batch::{LazyBatchColumn, LazyBatchColumnVec};
+    use crate::coprocessor::codec::data_type::VectorValue;
+    use crate::coprocessor::dag::expr::EvalConfig;
+    use cop_datatype::{EvalType, FieldTypeAccessor, FieldTypeTp};
+    use tipb::expression::FieldType;
+
+    struct MockExecutor {
+        // ID(INT,PK), Foo(INT), Bar(Float,Default 4.5)
+        data: Vec<(i64, Option<i64>, Option<f64>)>,
+        field_types: Vec<FieldType>,
+        offset: usize,
+        cfg: EvalConfig,
+    }
+
+    fn field_type(ft: FieldTypeTp) -> FieldType {
+        let mut f = FieldType::new();
+        f.as_mut_accessor().set_tp(ft);
+        f
+    }
+
+    impl MockExecutor {
+        /// create the MockExecutor with fixed schema and data.
+        fn new() -> MockExecutor {
+            let expect_rows = vec![
+                (1, Some(10), Some(5.2)),
+                (3, Some(-5), None),
+                (4, None, Some(4.5)),
+                (5, None, Some(0.1)),
+                (6, None, Some(4.5)),
+            ];
+
+            let field_types = vec![
+                field_type(FieldTypeTp::LongLong),
+                field_type(FieldTypeTp::LongLong),
+                field_type(FieldTypeTp::Double),
+            ];
+
+            MockExecutor {
+                data: expect_rows,
+                field_types,
+                offset: 0,
+                cfg: EvalConfig::default(),
+            }
+        }
+    }
+
+    impl BatchExecutor for MockExecutor {
+        #[inline]
+        fn next_batch(&mut self, expect_rows: usize) -> BatchExecuteResult {
```

Comment:
This mock executor cannot test this scenario:

```
next_batch() returns some data
next_batch() returns some data
next_batch() returns no data   <------ this may happen in selection executor
next_batch() returns some data
……
next_batch() returns no data and is drained
```

Reply:
I think we should consider the following dimensions:
    

- `next_batch()` returns ok
       -- get more than limit rows
       -- get exactly limit rows.
       -- get less rows
	
	

- `next_batch()` return's an error
		-- error happens before the `limit` rows
		-- error happens after the `limit` rows

And here `no data` belongs to `get less rows`, no matter is drained or not.

Reply:
That's fine but I still think 0 row is worth to test because it is a very special case that the underlying executor is allowed to return 0 columns in this case.

Reply:
um...I think this test case is a must tested for `TableScan` and `IndexScan`, not related to this PR.

Reply:
@AndreMouche Unit tests not just test existing code is correct, they also ensure that future change won't break something. So the test of zero rows will make sure that such stupid code in future won't pass.